### PR TITLE
chore(logplex source): refactor to use util/HttpSource

### PR DESCRIPTION
**Pull request by bill-bateman**
Refactoring the logplex source to reuse code from the http source.

Mainly, this moves the creation of the warp server out of the logplex code.

### CODE REVIEW

File: src/sources/logplex.rs [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
+use warp::filters::body::FullBody;
+use warp::http::{HeaderMap, StatusCode};
 
 #[derive(Deserialize, Serialize, Debug, Clone)]
 pub struct LogplexConfig {
     address: SocketAddr,
     tls: Option<TlsConfig>,
 }
 
+#[derive(Clone)]
```
Comment (Hoverbear) [inline src/sources/logplex.rs]:
```suggestion
#[derive(Clone, Default)]
```

File: src/sources/logplex.rs [hunk 2]
```diff
@@ -29,52 +43,8 @@ impl SourceConfig for LogplexConfig {
...
-
-        info!(message = "building logplex server", addr = %self.address);
-
-        let tls = TlsSettings::from_config(&self.tls, true)?;
-        let incoming = MaybeTlsIncoming::bind(&self.address, tls)?;
-
-        let server = warp::serve(routes).serve_incoming_with_graceful_shutdown(incoming, tripwire);
-
-        Ok(Box::new(server))
+        let source = LogplexSource {};
```
Comment (Hoverbear) [inline src/sources/logplex.rs]:
```suggestion
        let source = LogplexSource::default();
```

File: src/sources/logplex.rs [hunk 3]
```diff
@@ -86,7 +56,56 @@ impl SourceConfig for LogplexConfig {
     }
 }
 
-fn body_to_events(body: impl Buf) -> Vec<Event> {
+fn decode_message(body: FullBody, header_map: HeaderMap) -> Result<Vec<Event>, ErrorMessage> {
+    //deal with headers
```
Comment (Hoverbear) [inline src/sources/logplex.rs]:
```suggestion
    // Deal with headers
```

File: src/sources/logplex.rs [hunk 4]
```diff
@@ -86,7 +56,56 @@ impl SourceConfig for LogplexConfig {
...
+    let msg_count = match usize::from_str(get_header(&header_map, "Logplex-Msg-Count")?) {
+        Ok(v) => v,
+        Err(e) => return Err(header_error_message("Logplex-Msg-Count", &e.to_string())),
+    };
+    let frame_id = get_header(&header_map, "Logplex-Frame-Id")?;
+    let drain_token = get_header(&header_map, "Logplex-Drain-Token")?;
+    info!(message = "Handling logplex request", %msg_count, %frame_id, %drain_token);
+
+    //deal with body
```
Comment (Hoverbear) [inline src/sources/logplex.rs]:
```suggestion
    // Deal with body
```

File: src/sources/logplex.rs [hunk 5]
```diff
@@ -86,7 +56,56 @@ impl SourceConfig for LogplexConfig {
...
+    let events = body_to_events(body);
+
+    if events.len() != msg_count {
+        if cfg!(test) {
+            panic!("Parsed event count does not match message count header");
```
Comment (Hoverbear) [inline src/sources/logplex.rs]: Can we add the count information here too? I see it below!

We could probably just do the format once and let these share the message.

Comment (bill-bateman) [reply src/sources/logplex.rs]: Did this and the other suggestions in the new commit below
Thanks for the feedback 😄

**Pull request merged**